src/compiler/checker.ts

```diff
@@ -18083,22 +18084,27 @@ namespace ts {
                 }
                 expandingFlags = saveExpandingFlags;
                 depth--;
-                if (result) {
-                    if (result === Ternary.True || depth === 0) {
-                        if (result === Ternary.True || result === Ternary.Maybe) {
-                            // If result is definitely true, record all maybe keys as having succeeded. Also, record Ternary.Maybe
-                            // results as having succeeded once we reach depth 0, but never record Ternary.Unknown results.
-                            for (let i = maybeStart; i < maybeCount; i++) {
-                                relation.set(maybeKeys[i], RelationComparisonResult.Succeeded | propagatingVarianceFlags);
-                            }
+                if (result === Ternary.True || result === Ternary.False || depth === 0) {
+                    if (result === Ternary.True || result === Ternary.Maybe) {
+                        // If result is definitely true, record all maybe keys as having succeeded. Also, record Ternary.Maybe
+                        // results as having succeeded once we reach depth 0, but never record Ternary.Unknown results.
+                        for (let i = maybeStart; i < maybeCount; i++) {
+                            relation.set(maybeKeyList[i], RelationComparisonResult.Succeeded | propagatingVarianceFlags);
+                        }
+                    }
+                    else if (result === Ternary.False) {
+                        // A false result goes straight into global cache (when something is false under
+                        // assumptions it will also be false without assumptions)
+                        relation.set(id, (reportErrors ? RelationComparisonResult.Reported : 0) | RelationComparisonResult.Failed | propagatingVarianceFlags);
+                    }
+                    if (maybeStart === 0) {
+                        maybeKeySet.clear();
+                    }
+                    else {
+                        for (let i = maybeStart; i < maybeCount; i++) {
+                            maybeKeySet.delete(maybeKeyList[i]);

```

Comment:
What proportion of the set do we expect to delete?  How does deletion compare to clearing and rebuilding from `maybeKeyList`?

Reply:
It depends. Most of the time we just remove a single key, but in some degenerate cases (like the original issue) we end up removing a lot of keys at once. But it only happens a single time, so probably not worth optimizing beyond what's here now.